File: Same Tree.java
Given the roots of two binary trees p and q, write a function to check if they are the same or not.

Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

class Solution 
{
    public boolean isSameTree(TreeNode p, TreeNode q) 
    {
        // Base case: when both nodes are null, they must be equal. Assuming both root nodes are null.
        if(p == null && q == null) return true;
        
        // If either one of them is not null, but the other one is null. They cannot be identical.
        if(p == null && q != null) return false;
        if(p != null && q == null) return false;
        
        // If both nodes are not null, but their vals are different. They cannot be the same trees.
        if(p.val != q.val) return false;
        
        return (isSameTree(p.left,q.left) && isSameTree(p.right,q.right));
    }
}

https://leetcode.com/problems/same-tree/

File: Same Tree.py
Given the roots of two binary trees p and q, write a function to check if they are the same or not.

Two binary trees are considered the same if they are structurally identical, and the nodes have the same value.

Input: p = [1,2,3], q = [1,2,3]
Output: true

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        # First we check if nodes are Null/None
        # Three scenarios could happen:
        #    1: both are Null/None  ->  True
        #    2: Left is Null/None but Right isn't  ->  False
        #    3: Right is Null/None but Left isn't  ->  False
        if not p and not q:
            return True
        if not p or not q:
            return False
        
        # If the Null/None checks are good, we then check the values.
        if p.val != q.val:
            return False
        
        # We go check the left subtrees and right subtrees.
        return self.isSameTree(p.left, q.left) and self.isSameTree(p.right, q.right)
